File: server/utils/yamlResources.js
'use strict';

/* Builtin Modules */
var fs   = require('fs-extra');
var path = require('path');

/* 3rd-party Modules */
var yaml = require('js-yaml');

/* Project Modules */
var toolkit = require('./toolkit');

var FILE_CACHE = {};

/* Configure */
var CONFIG_KEY           = 'CONFIG';
var ENV_CONFIG_PREFIX    = 'DFF_';
var CUSTOM_CONFIG_PREFIX = 'CUSTOM_';

/**
 * Load a YAML file.
 *
 * @param  {String} key - Key of file data to load
 * @param  {String} filePath - Path of YAML file to load
 * @return {Object} Loaded file data
 */
var loadFile = exports.loadFile = function loadFile(key, filePath) {
  var fileContent = fs.readFileSync(filePath);
  var obj         = yaml.load(fileContent);

  if (key in FILE_CACHE) {
    Object.assign(FILE_CACHE[key], obj);
  } else {
    FILE_CACHE[key] = obj;
  }

  return obj;
};

/**
 * Load a YAML file as config file.
 *
 * @param {String}   configFilePath - Path of config file
 * @param {Function} calback
 */
var loadConfig = exports.loadConfig = function loadConfig(configFilePath, callback) {
  var configObj     = loadFile(CONFIG_KEY, configFilePath);
  var userConfigObj = {};

  // Collect config field type map
  var configTypeMap = {};
  for (var k in configObj) {
    switch(typeof configObj[k]) {
      case 'number':
        if (configObj[k].toString().match(/^\d+$/)) {
          configTypeMap[k] = 'integer';

        } else {
          configTypeMap[k] = 'float';
        }
        break;

      case 'boolean':
        configTypeMap[k] = 'boolean';
        break;

      default:
        if (toolkit.endsWith(k, '_LIST')) {
          configTypeMap[k] = 'list';

        } else if (toolkit.endsWith(k, '_MAP')) {
          configTypeMap[k] = 'map';

        } else {
          configTypeMap[k] = 'string';
        }
        break;
    }
  }

  var userConfigPath = process.env['CONFIG_FILE_PATH'] || configObj.CONFIG_FILE_PATH;
  if (!userConfigPath) {
    // User config path NOT SET
    console.log('[YAML Resource] ENV `CONFIG_FILE_PATH` not set. Use default config.');

  } else {
    // User config from FILE
    if (!fs.existsSync(userConfigPath)) {
      console.log(toolkit.strf('[YAML Resource] Config file `{0}` not found. Use default config.', userConfigPath));

    } else {
      var userConfigContent = fs.readFileSync(userConfigPath);
      userConfigObj = yaml.load(userConfigContent);

      Object.assign(configObj, userConfigObj)

      console.log(toolkit.strf('[YAML Resource] Config Overrided by: `{0}`', userConfigPath));
    }
  }

  // User config from env
  for (var envK in process.env) {
    if (!toolkit.startsWith(envK, ENV_CONFIG_PREFIX)) continue;

    var k = envK.slice(ENV_CONFIG_PREFIX.length);
    var v = process.env[envK];

    if ('string' === typeof v && v.trim() === '') {
      continue;
    }

    if (k in configObj) {
      configObj[k] = v;
      console.log(toolkit.strf('[YAML Resource] Config item `{0}` Overrided by env.', k));

    } else if (toolkit.startsWith(k, CUSTOM_CONFIG_PREFIX)) {
      configObj[k] = v;
      console.log(toolkit.strf('[YAML Resource] Custom config item `{0}` added by env.', k));
    }
  }

  // Convert config value type
  for (var k in configObj) {
    var v = configObj[k];
    var type = configTypeMap[k];

    if (!type) continue;

    // Set
    if (v === null) {
      switch(type) {
        case 'integer':
        case 'float':
          configObj[k] = 0;
          break;

        case 'list':
          configObj[k] = [];
          break;

        case 'map':
          configObj[k] = {};
          break;

        case 'string':
          configObj[k] = '';
          break;

        case 'boolean':
          configObj[k] = false;
          break;
      }
      continue;
    }

    switch(type) {
      case 'integer':
        configObj[k] = parseInt(v);
        break;

      case 'float':
        configObj[k] = parseFloat(v);
        break;

      case 'list':
        if (Array.isArray(v)) break;

        configObj[k] = v.toString();
        if (configObj[k].length > 0) {
          configObj[k] = v.trim().split(/[, \n]+/g).map(function(x) {
            return x.trim();
          });

        } else {
          configObj[k] = [];
        }
        break;

      case 'map':
        if ('object' === typeof v) break;

        var itemMap = {};
        v.split(',').forEach(function(item) {
          var itemParts = item.split('=');
          var itemK = itemParts[0].trim();
          var itemV = (itemParts[1] || '').trim();
          itemMap[itemK] = itemV;
        });
        configObj[k] = itemMap;
        break;

      case 'string':
        configObj[k] = v.toString();
        break;

      case 'boolean':
        configObj[k] = toolkit.toBoolean(v);
        break;
    }
  }

  if (CONFIG_KEY in FILE_CACHE) {
    Object.assign(FILE_CACHE[CONFIG_KEY], configObj);
  } else {
    FILE_CACHE[CONFIG_KEY] = configObj;
  }

  return callback(null, configObj, userConfigObj);
};

/**
 * Get the loaded YAML file data.
 *
 * @param  {String} key - Key of loaded data
 * @return {Object} Loaded file data
 */
exports.get = function get(key) {
  key = key.replace('.yaml', '');
  var resource = FILE_CACHE[key] || null;

  return resource;
};

/**
 * Set the value of loaded YAML file data.
 * @param {String} key
 * @param {String} path
 * @param {Any}    value
 */
exports.set = function set(key, path, value) {
  eval(toolkit.strf('FILE_CACHE["{0}"].{1} = {2}', key, path, JSON.stringify(value)));
};

/**
 * Get all the loaded YAML file data.
 *
 * @return {Object} All loaded file data
 */
exports.getAll = function getAll() {
  return FILE_CACHE;
};

File: worker/utils/yaml_resources.py
# -*- coding: utf-8 -*-

# Built-in Modules
import os
import sys

# Project Modules
from worker.utils import toolkit

# 3rd-party Modules
import six
import yaml
import requests

if six.PY2:
    FILE_OPEN_KWARGS = {}
else:
    FILE_OPEN_KWARGS = dict(encoding='utf8')

FILE_CACHE = {}

# Configure
CONFIG_KEY           = 'CONFIG'
ENV_CONFIG_PREFIX    = 'DFF_'
CUSTOM_CONFIG_PREFIX = 'CUSTOM_'
PRINT_DETAIL         = sys.argv[0].split('/')[-1] == 'celery'

def load_file(key, file_path):
    obj = None
    with open(file_path, 'r', **FILE_OPEN_KWARGS) as _f:
        file_content = _f.read()
        obj = yaml.safe_load(file_content)

    if key in FILE_CACHE:
        FILE_CACHE[key].update(obj)
    else:
        FILE_CACHE[key] = obj

    return obj

def load_config(config_file_path):
    config_obj = load_file(CONFIG_KEY, config_file_path)

    # Collect config field type map
    config_type_map = {}
    for k, v in config_obj.items():
        if v is True or v is False:
            config_type_map[k] = 'boolean'

        elif isinstance(v, int):
            config_type_map[k] = 'integer'

        elif isinstance(v, float):
            config_type_map[k] = 'float'

        else:
            if k.endswith('_LIST') or isinstance(v, (list, tuple)):
                config_type_map[k] = 'list'

            elif k.endswith('_MAP') or isinstance(v, dict):
                config_type_map[k] = 'map'

            else:
                config_type_map[k] = 'string'

    user_config_path = os.environ.get('CONFIG_FILE_PATH') or config_obj.get('CONFIG_FILE_PATH')
    if not user_config_path:
        # User config path NOT SET
        if PRINT_DETAIL:
            print('[YAML Resource] ENV `CONFIG_FILE_PATH` not set. Use default config')

    else:
        # User config from FILE
        if not os.path.exists(user_config_path):
            if PRINT_DETAIL:
                print('[YAML Resource] Config file `{}` not found. Use default config.'.format(user_config_path))

        else:
            user_config_obj = None
            with open(user_config_path, 'r', **FILE_OPEN_KWARGS) as _f:
                user_config_content = _f.read()
                user_config_obj = yaml.safe_load(user_config_content)

            config_obj.update(user_config_obj)

            if PRINT_DETAIL:
                print('[YAML Resource] Config Overrided by: `{}`'.format(user_config_path))

    # User config from env
    for env_k, v in os.environ.items():
        if not env_k.startswith(ENV_CONFIG_PREFIX):
            continue

        k = env_k[len(ENV_CONFIG_PREFIX):]

        if isinstance(v, str) and v.strip() == '':
            continue

        if k in config_obj:
            # Config override
            config_obj[k] = v
            if PRINT_DETAIL:
                print('[YAML Resource] Config item `{}` Overrided by env.'.format(k))

        elif k.startswith(CUSTOM_CONFIG_PREFIX):
            # Custom config
            config_obj[k] = v
            if PRINT_DETAIL:
                print('[YAML Resource] Custom config item `{}` added by env.'.format(k))

    # Convert config value type
    for k, v in config_obj.items():
        type_ = config_type_map.get(k)

        if not type_:
            continue

        if v is None:
            if type_ == 'boolean':
                config_obj[k] = False

            elif type_ == 'integer':
                config_obj[k] = 0

            elif type_ == 'float':
                config_obj[k] = 0.0

            elif type_ == 'list':
                config_obj[k] = []

            elif type_ == 'map':
                config_obj[k] = {}

            elif type_ == 'string':
                config_obj[k] = ''

            continue

        if type_ == 'boolean':
            config_obj[k] = toolkit.to_boolean(v)

        elif type_ == 'integer':
            config_obj[k] = int(v)

        elif type_ == 'float':
            config_obj[k] = float(v)

        elif type_ == 'list' and not isinstance(v, (tuple, list)):
            v = str(v)
            if len(v) > 0:
                config_obj[k] = list(map(lambda x: x.strip(), v.split(',')))
            else:
                config_obj[k] = []

        elif type_ == 'map' and not isinstance(v, dict):
            item_map = {}
            for item in v.split(','):
                item_parts = item.split('=')
                item_k = item_parts[0].strip()
                item_v = ''
                if len(item_parts) > 1:
                    item_v = item_parts[1].strip()
                item_map[item_k] = item_v

            config_obj[k] = item_map

        elif type_ == 'string':
            config_obj[k] = str(v)

    # Remap
    if config_obj.get('__REMAP'):
        for _from, _to in config_obj['__REMAP'].items():
            config_obj[_to] = config_obj.pop(_from, None)

    # Cache
    if CONFIG_KEY in FILE_CACHE:
        FILE_CACHE[CONFIG_KEY].update(config_obj)
    else:
        FILE_CACHE[CONFIG_KEY] = config_obj

    return config_obj

def get(key):
    key = key.replace('.yaml', '')
    resource = FILE_CACHE[key]

    return resource

def get_all():
    return FILE_CACHE
